### PR TITLE
Font loading

### PR DESCRIPTION
**Pull request by gestchild**
Attempting to balance performance, i.e. removing FOIT, preventing unnecessary downloads and limiting the impact of the FOUT.

Loading fonts with fontfaceloader - prevents FOIT - adds classes to html when fonts are loaded. These are then utilised in CSS to apply the fonts, replacing the fallback fonts.

When the fallback fonts are switched out in this way, it can be a quite jarring experience, especially for the Wellcome Bold font. Therefore, an inline subset of Wellcome bold font is loaded in the head.

SessionStorage is used to help optimise repeat views.

Most Mac/IOS devices have Helvetica Neue fonts available on the system. Therefore making use of 'local' in @font-face rules, to prevent unnecessary downloading of web fonts.

**Questions:**

- This obviously requires blocking CSS and javascript in the head. ~~I've set up a gulp task to generate includes from the js files which can then be called on the templates. Is there a better way to do this?~~
- ~~Using ES6 and ES6 modules for this head javascript inevitably adds bloat, should we just include the fontfaceloader inline? Any thoughts?~~ Not using ES6, to remove necessity of compiling the js

### CODE REVIEW

File: client/gulpfile.js [hunk 1]
```diff
@@ -28,6 +29,14 @@ const sources = {
     distPath: '../dist/assets/js/',
     all: 'js/**/*.js'
   },
+  jsHead: {
```
Comment (davidpmccormick) [inline client/gulpfile.js]: Would this make more sense as a (e.g. `head`) key on the `sources.js` object? Or maybe it does make more sense to keep them separate.

File: client/gulpfile.js [hunk 2]
```diff
@@ -37,6 +46,11 @@ const sources = {
   }
 };
 
+gulp.task('fonts:copy', function() {
+   gulp.src('./fonts/**/*.{woff,woff2}')
```
Comment (davidpmccormick) [inline client/gulpfile.js]:
```javascript
gulp.src(sources.fonts.srcPath)
  .pipe(gulp.dest(sources.fonts.distPath));
```
?

File: client/js/jsHead.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+const fontA = new FontFaceObserver('Wellcome Bold Web', {weight: 'bold'});
+const fontB = new FontFaceObserver('Helvetica Neue Light Web');
+const fontC = new FontFaceObserver('Helvetica Neue Medium Web');
+const fontD = new FontFaceObserver('Lettera Regular Web');
+const rootElement = document.documentElement;
+if (sessionStorage.fontsLoaded) {
+  document.documentElement.className += ' font-stage-1 font-stage-2';
+} else {
+  Promise.all([fontASubset.load()]).then(() => {
+    rootElement.className += ' font-stage-1';
```
Comment (davidpmccormick) [inline client/js/jsHead.js]: What is `rootElement` here? Should it be `document.documentElement`?

File: client/js/jsHead.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+/* eslint-env browser */
+import FontFaceObserver from 'fontfaceobserver';
+
+const fontASubset = new FontFaceObserver('Wellcome Bold Web Subset', {weight: 'bold'});
+const fontA = new FontFaceObserver('Wellcome Bold Web', {weight: 'bold'});
+const fontB = new FontFaceObserver('Helvetica Neue Light Web');
+const fontC = new FontFaceObserver('Helvetica Neue Medium Web');
+const fontD = new FontFaceObserver('Lettera Regular Web');
+const rootElement = document.documentElement;
+if (sessionStorage.fontsLoaded) {
```
Comment (davidpmccormick) [inline client/js/jsHead.js]: What will happen if the browser doesn't support `sessionStorage`?

**Comment by jamesgorrie** [comment]
> SessionStorage is used to help optimise repeat views.

Why not localStorage?

File: client/gulpfile.js [hunk 5]
```diff
@@ -28,6 +29,14 @@ const sources = {
     distPath: '../dist/assets/js/',
     all: 'js/**/*.js'
   },
+  jsHead: {
+    entry: './js/jsHead.js',
+    distPath: '../server/views/_partials/'
```
Comment (jamesgorrie) [inline client/gulpfile.js]: We should never be building into the server folder - could we have this into the dist folder?

File: client/package.json [hunk 6]
```diff
@@ -7,9 +7,11 @@
   "devDependencies": {
     "babel-core": "^6.17.0",
     "babel-loader": "^6.2.5",
+    "babel-polyfill": "^6.16.0",
     "babel-preset-es2015": "^6.16.0",
     "browser-sync": "^2.17.0",
     "eslint-config-wellcome": "^1.0.8",
+    "fontfaceobserver": "^2.0.5",
```
Comment (jamesgorrie) [inline client/package.json]: Given that this is actually included in the app - should it not be in `deps`?

File: server/views/layout/_preview.njk [hunk 7]
```diff
@@ -5,9 +5,13 @@
...
   <title>Wellcome Collection</title>
   <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
+  <style>
+	 {% include "views/partials/css-head.njk" %}
+  </style>
   <link rel="stylesheet" href="{{ '/assets/css/styleguide.css' | path }}">
-  {# TODO Temporary until decide on font hosting solution #}
-  <script type="text/javascript" src="//fast.fonts.net/jsapi/b026248c-28d7-40e1-b542-97c76e7a74fd.js"></script>
+  <script>
+	 {% include "views/partials/js-head.njk" %}
```
Comment (jamesgorrie) [inline server/views/layout/_preview.njk]: This seems like a bit of a hack and were committing compiled code here.
I'm not 100% sure what the solution is - but could we at least have a `TODO` to get rid of this later?

File: client/gulpfile.js (hunk 2 above)
Comment (gestchild) [inline client/gulpfile.js]: Will update

File: client/js/jsHead.js (hunk 4 above)
Comment (gestchild) [inline client/js/jsHead.js]: At the moment error. We need to introduce a cutting the mustard test for all the javascript

File: client/gulpfile.js (hunk 5 above)
Comment (gestchild) [inline client/gulpfile.js]: Don't see why not, I'll move it.

Is it weird having nunjucks includes in the dist folder though? But I'm sticking in a TODO to revisit how we include inline js anyway.

File: client/package.json (hunk 6 above)
Comment (gestchild) [inline client/package.json]: It should, will move

**Comment by gestchild** [comment]
@jamesgorrie re sessionStorage. Web storage is being used to set a token to say the font has loaded (not to actually store the fonts). And so the subsequent font loads, with the token set, are relying on browser cache.

**Comment by jamesgorrie** [comment]
@gestchild so if I clear my localStorage, it won't make a request to the server thus giving me a broken font?

**Pull request merged**